File: android/app/src/main/java/voidawt/image/PixelGrabber.java
package voidawt.image;

import voidawt.Image;

public class PixelGrabber {
    private final Image image;
    private final int x;
    private final int y;
    private final int w;
    private final int h;
    private final int[] pix;
    private final int off;
    private final int scansize;

    public PixelGrabber(Image image, int x, int y, int w, int h, int[] pix, int off, int scansize) {
        this.image = image;
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
        this.pix = pix;
        this.off = off;
        this.scansize = scansize;
    }

    public boolean grabPixels() throws InterruptedException {
        int[] src = image.peekArgb();
        int iw = image.getWidth(null);
        int ih = image.getHeight(null);
        if (src == null || iw <= 0) {
            return false;
        }
        for (int row = 0; row < h; row++) {
            int sy = y + row;
            if (sy < 0 || sy >= ih) {
                continue;
            }
            for (int col = 0; col < w; col++) {
                int sx = x + col;
                if (sx < 0 || sx >= iw) {
                    continue;
                }
                pix[off + row * scansize + col] = src[sy * iw + sx];
            }
        }
        return true;
    }
}

File: src/commonMain/kotlin/awt/image/PixelGrabber.kt
package awt.image

import awt.Image

expect open class PixelGrabber {
    constructor(img: Image, x: Int, y: Int, w: Int, h: Int, pix: IntArray, off: Int, scansize: Int)
    constructor(ip: ImageProducer, x: Int, y: Int, w: Int, h: Int, pix: IntArray, off: Int, scansize: Int)
    constructor(img: Image, x: Int, y: Int, w: Int, h: Int, forceRGB: Boolean)

    fun startGrabbing()
    fun abortGrabbing()
    fun grabPixels(): Boolean
    fun grabPixels(ms: Long): Boolean
    fun getStatus(): Int
    fun getWidth(): Int
    fun getHeight(): Int
    fun getPixels(): Any?
    fun getColorModel(): ColorModel
}

expect val IMAGE_OBSERVER_FRAMEBITS: Int
expect val IMAGE_OBSERVER_ALLBITS: Int
expect val IMAGE_OBSERVER_ERROR: Int
expect val IMAGE_OBSERVER_ABORT: Int